### ISSUE TITLE
What about presets ?

### ISSUE THREAD

**Issue by nay-seven** [VCVRack/Rack#78]
I know you try to reproduce the most of hardware world, but as we talk about a software, do you plan to add a preset engine ( we can recall by MIDI for example) ?

**Comment by AndrewBelt** [comment]
You're suggesting adding a quick-save/load feature for single modules? Sounds like some cool features can come out of this idea.

**Comment by nay-seven** [comment]
yes,  each modules can have different presets  but also a global one, this way we can make evolve a patch (a smooth transition with moving knobs would be fantastic) ;-)

**Comment by sanderbaan** [comment]
And would be great if we can morph between those presets also :) like you can in Reaktor

**Comment by AndrewBelt** [comment]
Can someone convince me why we need module-level presets? Implementing it would be easy, but IMO it's easier and more straightforward to turn a few knobs rather than to right click and sift through a few presets.

**Comment by Uriel-A** [comment]
I don't see a need for individual module presets but I do have a need for partial presets, as in a group of modules or sub patch that can be loaded into the main patch.

It would speed things up a lot for stuffs like VCO-ADSR-VCA etc. especially when you want to have a few of them all ready patched up together and you just need to connect the remaining pitch, gate and output to the rest of your patch.

**Comment by sanderbaan** [comment]
I rather see something like snapshots where you can save states of all modules in the patch and recall them quickly and eventually can morph between them like you can in reaktor

**Comment by odahan** [comment]
I think the simplest way is to allow the loading of another patch within a patch. Thus you can reuse any patch as a part of another. It is the same concept as "presets" but it does not need a lot of modifications of the program. The loading sequence already exists. The only problem to solve (but it is easy) is the position of the new  modules. When loading a patch within another patch the loading sequence just needs to find the first free square space for the additional patch and apply an offset on x,y position of each modules. The user can then move them around to rearrange if needed.
This is leading to the need for a way to move multiple modules at the same time to ease rearrangement but I know this is another question.

**Comment by Uriel-A** [comment]
@odahan You're right. I overlooked the part when I'm gonna need to move the group of modules that I'll be loading into the main patch. There's a ongoing discussion regarding this over at https://github.com/VCVRack/Rack/issues/307

**Comment by michaelpalumbo** [comment]
another route could be for each module to expose its current state in terms of parameter-value(s), and its patch connections between ins/outs, or if this is already the case, have access to the database that is storing the current state of each module, and then build some module(s) around interacting with this database in order to do the morphing that was mentioned by @sanderbaan for example... thoughts? Given some info I'd be happy to contribute to this...

**Comment by Coirt** [comment]
@michaelpalumbo the autosave/patch file has positions of parameters I think (open with IDE/notepad) But if edited it would require a restart or reopening of the patch.

**Comment by tuborgman** [comment]
Presets could pretty much make sense with storing/loading sequencers data.
